app/src/main/java/org/p2p/wallet/history/ui/history/adapter/HistoryAdapter.kt

```diff
@@ -1,123 +1,125 @@
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HistoryTransactionViewHolder {
+        return when (viewType) {
+            TRANSACTION_VIEW_TYPE -> TransactionViewHolder(parent, onTransactionClicked)
+            HISTORY_EMPTY_VIEW_TYPE -> EmptyViewHolder(parent)
+            HISTORY_DATE_VIEW_TYPE -> DateViewHolder(parent)
+            PROGRESS_VIEW_TYPE -> ProgressViewHolder(parent)
             else -> ErrorViewHolder(parent)
         }
+    }
 
-    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+    override fun onBindViewHolder(holder: HistoryTransactionViewHolder, position: Int) {
         when (holder) {
-            is TransactionViewHolder -> holder.onBind(data[position] as HistoryItem.TransactionItem)
-            is DateViewHolder -> holder.onBind(data[position] as HistoryItem.DateItem)
-            is ErrorViewHolder -> {
-                val item = pagingState as? PagingState.Error
-                if (item != null) {
-                    holder.onBind(item.e.message.orEmpty(), onRetryClicked)
-                } else {
-                    holder.onBind(R.string.error_general_message, onRetryClicked)
-                }
-            }
-            is ProgressViewHolder -> {
-                // do nothing
-            }
-            is EmptyViewHolder -> {
-                // do nothing
-            }
+            is TransactionViewHolder -> holder.onBind(currentItems[position] as HistoryItem.TransactionItem)
+            is DateViewHolder -> holder.onBind(currentItems[position] as HistoryItem.DateItem)
+            is ErrorViewHolder -> holder.onBind(pagingController.currentPagingState, onRetryClicked)
+            else -> NoOp
         }
     }
 
     override fun getItemId(position: Int): Long {
-        return when (val item = if (position < data.size) data[position] else null) {
+        return when (val item = currentItems.getOrNull(position)) {
             is HistoryItem.TransactionItem -> item.transaction.signature.hashCode().toLong()
             is HistoryItem.DateItem -> item.date.hashCode().toLong()
             is HistoryItem.Empty -> position.hashCode().toLong()
             else -> RecyclerView.NO_ID
         }
     }
 
-    override fun getItemViewType(position: Int): Int = when {
-        !stateRequiresExtraItem(pagingState) || position < itemCount - 1 -> getAdapterViewType(position)
-        pagingState is PagingState.Loading || pagingState is PagingState.InitialLoading -> R.layout.item_progress
-        else -> R.layout.item_error
+    override fun getItemViewType(position: Int): Int {
+        val preLastItemPosition = itemCount - 1
+
+        // todo: refactor to AdapterDelegate
+        //  and create item list with progress item inside instead of implicitly adding it here
+        return when {
+            !pagingController.stateRequiresLoadingItem() || position < preLastItemPosition -> getAdapterViewType(
+                position
+            )
```

Comment:
If it exceeds max line, mb it'd be better to move the whole function to new line? 
I remeber we don't wrap lines like that